### PR TITLE
coprocessor, dag: add json functions.

### PR DESCRIPTION
**Pull request by hicqu**

### CODE REVIEW

File: src/coprocessor/dag/expr/fncall.rs [hunk 1]
```diff
@@ -176,6 +178,8 @@ impl FnCall {
             ScalarFuncSig::IfTime |
             ScalarFuncSig::IfDuration => (3, 3),
 
+            ScalarFuncSig::JsonArraySig | ScalarFuncSig::JsonObjectSig => (0, usize::MAX),
```
Comment (AndreMouche) [inline src/coprocessor/dag/expr/fncall.rs]: It seems inconsistent with `tidb` https://github.com/pingcap/tidb/blob/master/expression/builtin.go#L958

Comment (hicqu) [reply src/coprocessor/dag/expr/fncall.rs]: It's a TiDB bug. I will fix that.

File: src/coprocessor/dag/expr/json.rs [hunk 2]
```diff
@@ -0,0 +1,438 @@
...
+    pub fn json_object<'a, 'b: 'a>(
+        &'b self,
+        ctx: &StatementContext,
+        row: &'a [Datum],
+    ) -> Result<Option<Cow<'a, Json>>> {
+        let mut pairs = BTreeMap::new();
+        if !self.children.is_empty() {
+            let parser = JsonFuncArgsParser::new(ctx, row);
+            let keys = try_opt!(parser.get_strings(self.children.iter().step_by(2)));
+            let elems = try_opt!(parser.get_jsons(self.children[1..].iter().step_by(2), true));
```
Comment (BusyJay) [inline src/coprocessor/dag/expr/json.rs]: What if `keys.len() != elems.len()`?

Comment (hicqu) [reply src/coprocessor/dag/expr/json.rs]: We check this in `Json::extract`.

**Comment by hicqu** [comment]
@BusyJay , PTAL.

File: src/coprocessor/dag/expr/json.rs [hunk 3]
```diff
@@ -0,0 +1,442 @@
...
+            if keys.len() != elems.len() {
+                return Err(box_err!("Incorrect parameter count for 'json_object'"));
```
Comment (AndreMouche) [inline src/coprocessor/dag/expr/json.rs]: Could we check it in build?

File: src/coprocessor/dag/expr/json.rs (hunk 2 above)
Comment (hicqu) [reply src/coprocessor/dag/expr/json.rs]: Now we check that when build the tree.

**Comment by hicqu** [comment]
@BusyJay , PTAL again, thanks.

File: src/coprocessor/dag/expr/json.rs [hunk 4]
```diff
@@ -0,0 +1,455 @@
...
+        JsonFuncArgsParser { ctx: ctx, row: row }
+    }
+
+    #[inline]
+    fn parse<'b: 'a, It, T, F>(args: It, f: F) -> Result<Option<Vec<T>>>
+    where
+        It: Iterator<Item = &'b Expression>,
+        F: Fn(&'b Expression) -> Result<Option<T>>,
+    {
+        args.map(f).collect::<Result<Option<Vec<_>>>>()
```
Comment (BusyJay) [inline src/coprocessor/dag/expr/json.rs]: I think `args.map(f).collect()` is enough.

File: src/coprocessor/dag/expr/json.rs [hunk 5]
```diff
@@ -0,0 +1,455 @@
...
+                let j = try_opt!(e.eval_json(self.ctx, self.row)).into_owned();
+                Ok(Some(j))
+            };
+            JsonFuncArgsParser::parse(args, func)
+        } else {
+            let func = |e: &'b Expression| {
+                let j = try!(e.eval_json(self.ctx, self.row))
+                    .map(Cow::into_owned)
+                    .unwrap_or(Json::None);
+                Ok(Some(j))
```
Comment (BusyJay) [inline src/coprocessor/dag/expr/json.rs]: This feels more like nullable to me.

**Pull request merged**